lib/cretonne/meta/cdsl/ast.py

```diff
@@ -332,14 +368,51 @@ def inst_predicate(self):
         return pred
 
     def copy(self, m):
-        # type: (Dict[Var, Var]) -> Apply
+        # type: (VarMap) -> Apply
         """
         Return a copy of this Expr with vars replaced with fresh variables,
         in accordance with the map m. Update m as neccessary.
         """
         return Apply(self.inst, tuple(map(lambda e: replace_var(e, m),
                                           self.args)))
 
+    def vars(self):
+        # type: () -> Set[Var]
+        """ Return the set of all Vars that appear in self"""
+        res = set()
+        for i in self.inst.value_opnums:
+            arg = self.args[i]
+            assert isinstance(arg, Var)
+            res.add(arg)
+        return res
+
+    def substitution(self, other, s):
+        # type: (Apply, VarMap) -> Optional[VarMap]
+        """
+        If the application self and other agree structurally, return a variable
+        substitution to transform self ot other. Two applications agree
+        structurally if:
+            1) They are over the same instruction
+            2) Every Var v in self, maps to a single Var w in other. I.e for
+               each use of v in self, w is used in the corresponding place in
+               other.
+        """
+        if self.inst != other.inst:
+            return None
+
+        # TODO: Should we check imm/cond codes here as well?
```

Comment:
Probably, yes. Immediate operands (including condition codes) can be either Vars or literals.

Operands can also be entity references, most commonly a destination EBB for a branch.